arrow-buffer/src/builder/offset.rs

```diff
@@ -0,0 +1,238 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::ops::{Add, Sub};
+
+use crate::{ArrowNativeType, OffsetBuffer, ScalarBuffer};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OffsetsBuilder<O: ArrowNativeType> {
+    offsets: Vec<O>,
+}
+
+/// builder for [`OffsetBuffer`]
+impl<O: ArrowNativeType + Add<Output = O> + Sub<Output = O>> OffsetsBuilder<O> {
+    /// create a new builder containing only 1 zero offset
+    pub fn new(capacity: usize) -> Self {
+        let mut offsets = Vec::with_capacity(capacity + 1);
+        offsets.push(O::usize_as(0));
+        Self::new_unchecked(offsets)
+    }
+
+    /// create a new builder containing capacity number of zero offsets
+    pub fn new_zeroed(capacity: usize) -> Self {
+        let offsets = vec![O::usize_as(0); capacity + 1];
+        Self::new_unchecked(offsets)
+    }
+
+    /// create from offsets
+    /// caller guarantees that offsets are monotonically increasing values
+    #[inline]
+    pub fn new_unchecked(offsets: Vec<O>) -> Self {
```

Comment:
```suggestion
    pub unsafe fn new_unchecked(offsets: Vec<O>) -> Self {
```

Presuming we change to using unchecked in finish